cudamat/cudamat.py

```diff
@@ -1501,6 +1503,25 @@ def where(condition_mat, if_mat, else_mat, target=None):
     return target
 
 
+def correlate(mat, kernel, target=None):
+    """
+    Cross-correlate a matrix with a kernel matrix.
+    The kernel matrix is centered over each element of the matrix mat.
+    Width and height of the kernel matrix must be an odd integer.
+    """
+    if not target:
+        m = _cudamat.get_leading_dimension(mat.p_mat)
+        n = _cudamat.get_nonleading_dimension(mat.p_mat)
+
+        target = CUDAMatrix(np.empty((m, n)))
```

Comment:
This should be:
```python
target = empty((m, n))
```
Otherwise it will create an array on CPU and copy it to GPU. This incurs a lot of overhead:
```
%timeit cmt.correlate(X, Y, None)
100 loops, best of 3: 12.2 ms per loop
%timeit cmt.correlate(X, Y, cmt.empty(X.shape))
100 loops, best of 3: 6.83 ms per loop
```